File: plugins/claude-code/safe-change/hooks/lib/detect.py
"""dbt model file detection. Phase 1: dbt only. Extensible for future SQL frameworks."""
import os

EXCLUDED_DIRS = {"seeds", "analyses", "macros", "snapshots"}
DBT_EXTENSIONS = {".sql", ".yml", ".yaml"}
REF_PATTERNS = ["{{ ref(", "{{ source("]
MAX_SCAN_LINES = 50


def is_dbt_model(file_path: str) -> bool:
    """Check if file_path is a dbt model SQL file.

    Rules:
    - Extension is .sql only (yml/yaml schema files are tracked by PostToolUse
      but not gated by PreToolUse — they don't map to a single table)
    - Path contains /models/
    - NOT in seeds/, analyses/, macros/, snapshots/
    - Contains {{ ref( or {{ source( in first 50 lines, OR file doesn't exist
      yet (new model being created)
    """
    ext = os.path.splitext(file_path)[1].lower()
    if ext != ".sql":
        return False

    # Normalize path separators
    normalized = file_path.replace("\\", "/")

    if "/models/" not in normalized:
        return False

    # Check excluded directories
    parts = normalized.split("/")
    for part in parts:
        if part in EXCLUDED_DIRS:
            return False

    # If file doesn't exist yet, it's a new model being created — treat as dbt model
    if not os.path.exists(file_path):
        return True

    # Existing .sql files must contain ref() or source() in first 50 lines
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            for i, line in enumerate(f):
                if i >= MAX_SCAN_LINES:
                    break
                for pattern in REF_PATTERNS:
                    if pattern in line:
                        return True
    except (OSError, UnicodeDecodeError):
        return False

    return False


def is_dbt_schema_file(file_path: str) -> bool:
    """Check if file_path is a dbt schema yml file under models/.

    Used by PostToolUse for tracking edits, but not for PreToolUse gating.
    """
    ext = os.path.splitext(file_path)[1].lower()
    if ext not in (".yml", ".yaml"):
        return False

    normalized = file_path.replace("\\", "/")
    if "/models/" not in normalized:
        return False

    parts = normalized.split("/")
    for part in parts:
        if part in EXCLUDED_DIRS:
            return False

    return True


def extract_table_name(file_path: str) -> str:
    """Extract table name from file path.

    Example: /project/models/staging/client_hub_master.sql -> client_hub_master
    """
    basename = os.path.basename(file_path)
    name, _ = os.path.splitext(basename)
    return name

File: plugins/opencode/prevent/src/detect.ts
/**
 * dbt SQL file detection. Reads dbt_project.yml paths (cached) instead of
 * hardcoding folder names.
 *
 * Port of plugins/claude-code/prevent/hooks/lib/detect.py
 */
import { readFileSync, existsSync, statSync } from "fs";
import { basename, extname, dirname, join, resolve } from "path";
import { getDbtPaths } from "./cache";

const REF_PATTERNS = ["{{ ref(", "{{ source("];
const DBT_BLOCK_PATTERNS = ["{% macro", "{% snapshot"];
const MAX_SCAN_LINES = 50;

function getGatedDirs(filePath: string): Set<string> {
  const paths = getDbtPaths(filePath);
  const gated = new Set<string>();
  for (const key of ["model-paths", "macro-paths", "snapshot-paths"] as const) {
    for (const p of paths[key] ?? []) {
      const last = p.replace(/\/+$/, "").split("/").pop()!;
      gated.add(last);
    }
  }
  return gated;
}

function getExcludedDirs(filePath: string): Set<string> {
  const paths = getDbtPaths(filePath);
  const excluded = new Set<string>();
  for (const key of ["seed-paths", "analysis-paths"] as const) {
    for (const p of paths[key] ?? []) {
      const last = p.replace(/\/+$/, "").split("/").pop()!;
      excluded.add(last);
    }
  }
  return excluded;
}

function getMacroDirs(filePath: string): Set<string> {
  const paths = getDbtPaths(filePath);
  return new Set(
    (paths["macro-paths"] ?? []).map((p) =>
      p.replace(/\/+$/, "").split("/").pop()!
    )
  );
}

function setsIntersect(a: Set<string>, b: Set<string>): boolean {
  for (const item of a) {
    if (b.has(item)) return true;
  }
  return false;
}

/**
 * Check if filePath is a dbt SQL file that could affect data.
 *
 * Gates any .sql file in a dbt project (models, macros, snapshots) since
 * macros are inlined into models at compile time and can have equal or
 * greater blast radius than a single model change.
 */
export function isDbtModel(filePath: string): boolean {
  if (extname(filePath).toLowerCase() !== ".sql") return false;

  const normalized = filePath.replace(/\\/g, "/");
  const parts = new Set(normalized.split("/"));

  const excluded = getExcludedDirs(filePath);
  if (setsIntersect(parts, excluded)) return false;

  const gated = getGatedDirs(filePath);
  if (!setsIntersect(parts, gated)) return false;

  // New file being created — treat as dbt SQL
  if (!existsSync(filePath)) return true;

  // Existing .sql files must contain dbt patterns in first N lines
  try {
    const content = readFileSync(filePath, "utf-8");
    const lines = content.split("\n").slice(0, MAX_SCAN_LINES);
    for (const line of lines) {
      for (const pattern of REF_PATTERNS) {
        if (line.includes(pattern)) return true;
      }
      for (const pattern of DBT_BLOCK_PATTERNS) {
        if (line.includes(pattern)) return true;
      }
    }
  } catch {
    return false;
  }

  return false;
}

/**
 * Check if filePath is a dbt schema yml file under a model directory.
 * Used by post-edit tracking, but not for pre-edit gating.
 */
export function isDbtSchemaFile(filePath: string): boolean {
  const ext = extname(filePath).toLowerCase();
  if (ext !== ".yml" && ext !== ".yaml") return false;

  const normalized = filePath.replace(/\\/g, "/");
  const parts = new Set(normalized.split("/"));

  const excluded = getExcludedDirs(filePath);
  if (setsIntersect(parts, excluded)) return false;

  const paths = getDbtPaths(filePath);
  const modelDirs = new Set(
    (paths["model-paths"] ?? []).map((p) =>
      p.replace(/\/+$/, "").split("/").pop()!
    )
  );
  return setsIntersect(parts, modelDirs);
}

/**
 * Check if filePath is under a macro directory.
 */
export function isMacroFile(filePath: string): boolean {
  const normalized = filePath.replace(/\\/g, "/");
  const parts = new Set(normalized.split("/"));
  return setsIntersect(parts, getMacroDirs(filePath));
}

/**
 * Extract table or macro name from file path.
 *
 * Example: /project/models/staging/client_hub_master.sql -> client_hub_master
 * Example: /project/macros/filters/internal_accounts.sql -> macro:internal_accounts
 */
export function extractTableName(filePath: string): string {
  const base = basename(filePath);
  const name = base.slice(0, base.length - extname(base).length);
  if (isMacroFile(filePath)) return `macro:${name}`;
  return name;
}
